hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java

```diff
@@ -1511,24 +1512,41 @@ private void processNewState(Collection<MigrationInfo> allCompletedMigrations,
             lock.lock();
             try {
                 processMigrations(allCompletedMigrations, allActiveMigrations);
+                Version version = node.getClusterService().getClusterVersion();
+                boolean version312plus = version.isGreaterOrEqual(Versions.V3_12);
 
                 if (newState != null) {
-                    maxVersion = Math.max(maxVersion, getPartitionStateVersion()) + 1;
-                    logger.info("Applying the most recent of partition state...");
+                    maxVersion = Math.max(maxVersion, getPartitionStateVersion());
+                    // RU_COMPAT_3_11
+                    if (version312plus) {
```

Comment:
A few clarifications:
- since it's purely the master here determining the p-table version, would it be possible if we used the same logic regardless if the cluster version is 3.11 or 3.12? Or is there some issue if a 3.12 member sets a too-high version while running in a 3.11 cluster? And if there is an issue with setting a too-high version, is it ok to increment the version regardless of the cluster version (after `Increment version once more to make it greater than the most recent version.`)?
- from what I understand in the issue this PR fixes, the `Partition state versions are not set` when migration info's are sent from a 3.11 to a 3.12 member which becomes master. But I don't see us setting the migration info state versions here in the PR. We just set the increment in `FetchMostRecentPartitionTableTask#processMigrations`. Am I missing something?

Reply:
Thanks for the review!

- Setting partition table version to a higher value doesn't break anything. We want just to make sure that when master changes and decides the most recent partition table, it increments version to a bigger value than ever seen in the cluster. But in a 3.11 cluster, we don't have version information per-migration, that's why `migration.getFinalPartitionVersion()` cannot be used. Even it throws `IllegalStateException` to  make it clear, when  partition version is not set (yet), it fails.

-  Issue happens when cluster version is upgraded to 3.12 and master leaves when cluster is 3.12. 
Every member keeps a list of recently completed migrations. Master evicts completed migrations from that list when all members learn that migration. After master evicts them, other members evict from their lists when they receive an update from master. So at any time, recently completed migrations list contain some elements. 
In a 3.11 cluster, those migrations do not have version information but when upgraded to 3.12, partition versions are needed. But... problem is, we cannot generate/calculate those versions during upgrade. Because we don't have exact knowledge. 
Good thing is, we don't have to :) Cluster version upgrade is a safepoint for migrations, we exactly know that all members have the same and most recent partition table. So, we can just evict all completed migrations from list during version upgrade. See `MigrationManager.onClusterVersionChange(newVersion)`.

Reply:
Got it, thanks! Merge time! 💯